File: ark_oauth_client_nodejs/examples/client-credentials.js
/**
 * Service-to-service: a job or daemon authenticating as itself, with no user involved.
 *
 *   ARK_CLIENT_ID=my-service ARK_CLIENT_SECRET=… node examples/client-credentials.js
 *
 * The token this returns carries the *service's* authority, not a person's. Never use it to act
 * on behalf of a signed-in user: nothing downstream can tell the difference, and the audit trail
 * says the service did it.
 */
import { ArkOAuthClient } from '../src/index.js';
import { authority, clientId, clientSecret } from './config.js';

const client = new ArkOAuthClient({ authority, clientId, clientSecret });

// Cached until shortly before it expires — asking for a token per outbound call turns one request
// into two and rate-limits the service against its own identity provider.
const token = await client.clientCredentials({ scopes: ['reports.read'] });

console.log('access token :', `${token.accessToken.slice(0, 32)}…`);
console.log('expires in   :', token.expiresIn(), 'seconds');
console.log('scopes       :', token.scopes().join(' ') || '(none)');
console.log('subject      :', token.subject, '(the client itself)');

const again = await client.clientCredentials({ scopes: ['reports.read'] });
console.log('cached       :', again.accessToken === token.accessToken);

const response = await fetch(`${authority}/oauth2/userinfo`, {
  headers: { Authorization: token.authorizationHeader() }
});
console.log('userinfo     :', response.status, '(403 is correct — there is no user in this flow)');

File: ark_oauth_client_py/examples/client_credentials.py
"""
The client credentials grant — a service authenticating as itself, with no user involved.

    ARK_AUTHORITY=... ARK_CLIENT_ID=my-service ARK_CLIENT_SECRET=... python examples/client_credentials.py

Never use the resulting token to act on behalf of a signed-in user: it carries the service's
authority rather than theirs, and nothing downstream can tell the difference.
"""

import sys

from ark_oauth_client import ArkOAuthClient

from config import AUTHORITY, CLIENT_ID, CLIENT_SECRET, REQUIRE_HTTPS


def main():
    if not CLIENT_SECRET:
        sys.exit("set ARK_CLIENT_SECRET — a public client cannot use this grant.")

    client = ArkOAuthClient(
        authority=AUTHORITY,
        client_id=CLIENT_ID,
        client_secret=CLIENT_SECRET,
        require_https=REQUIRE_HTTPS,
    )

    tokens = client.client_credentials(scopes=["reports.read"])
    print(f"access token : {tokens.access_token[:32]}...")
    print(f"expires in   : {tokens.expires_in()}s")
    print(f"scopes       : {', '.join(tokens.scopes()) or '(none granted)'}")

    # Cached until shortly before expiry: asking again costs nothing.
    again = client.client_credentials(scopes=["reports.read"])
    print(f"cached       : {again is tokens}")


if __name__ == "__main__":
    main()
